### ISSUE TITLE
Cache mp3 with encryption throws ParseException.

### ISSUE THREAD

**Issue by Toffor** [google/ExoPlayer#3883]
First of all thanks for this great library. I am using version 2.6.0 and tried to implement cache with encryption while listening media. I tried the example code in here https://github.com/google/ExoPlayer/issues/2467#issuecomment-281119679
And my code like below

```
 val simpleExoPlayer : SimpleExoPlayer by lazy {
        val bandwidthMeter = DefaultBandwidthMeter()
        val trackSelectionFactory = AdaptiveTrackSelection.Factory(bandwidthMeter)
        val trackSelector = DefaultTrackSelector(trackSelectionFactory)
        ExoPlayerFactory.newSimpleInstance(context, trackSelector)!!
    }

    fun prepareMediaSource(url: String) {
        simpleExoPlayer.prepare(createMediaSource(url))
    }

    private fun createMediaSource(url: String) = ExtractorMediaSource(Uri.parse(url), cacheDataSourceFactory(), DefaultExtractorsFactory(), null, null)

    private fun cacheDataSourceFactory() = DataSource.Factory {
        val cache = SimpleCache(cacheDir(), NoOpCacheEvictor())
        val secret = generateSecret("password")
        val aesCipherDataSource = AesCipherDataSource(secret, FileDataSource())
        val aesCipherDataSink = AesCipherDataSink(secret, CacheDataSink(cache, Long.MAX_VALUE))
        CacheDataSource(cache, okHttpDataSourceFactory().createDataSource(), aesCipherDataSource, aesCipherDataSink, CacheDataSource.FLAG_BLOCK_ON_CACHE, null)
    }

    private fun okHttpDataSourceFactory() = OkHttpDataSourceFactory(OkHttpClient(), Util.getUserAgent(context.applicationContext, context.packageName), null)

    private fun cacheDir() = context.externalCacheDir

    private fun generateSecret(key: String) : ByteArray{
        val keyBytes = ByteArray(16)
        val md = MessageDigest.getInstance("SHA-256")
        md.update(key.toByteArray())
        System.arraycopy(md.digest(), 0, keyBytes, 0, keyBytes.size)
        return keyBytes
    }
```

And this is sample mp3 
https://archive.org/download/Mp3Playlist_555/AaronNeville-CrazyLove.mp3

If i use okHttpDataSourceFactory() instead of cacheDataSourceFactory() when creating ExtractorMediaSource everything goes well. But if i try to use cacheDataSource sometimes its working well but mostly it throws the exception below

 E/ExoPlayerImplInternal: Source error.
                                                                                         com.google.android.exoplayer2.ParserException: Searched too many bytes.
                                                                                             at com.google.android.exoplayer2.extractor.mp3.Mp3Extractor.synchronize(Mp3Extractor.java:265)
                                                                                             at com.google.android.exoplayer2.extractor.mp3.Mp3Extractor.read(Mp3Extractor.java:172)
                                                                                             at com.google.android.exoplayer2.source.ExtractorMediaPeriod$ExtractingLoadable.load(ExtractorMediaPeriod.java:714)
                                                                                             at com.google.android.exoplayer2.upstream.Loader$LoadTask.run(Loader.java:315)
                                                                                             at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1162)
                                                                                             at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:636)
                                                                                             at java.lang.Thread.run(Thread.java:764)

How can i fix this? Thanks.

**Comment by andrewlewis** [comment]
It looks a new `SimpleCache` is being created every time `createDataSource` is called, whereas the linked issue uses the same instance on each invocation.

**Comment by Toffor** [comment]
Thanks for response, i did simple cache final and deleted app from emulator.

`    private val simpleCache = SimpleCache(cacheDir(), cacheEvictor)
`
When ı run app, first i take this error.
-----------------------------------------------------------------------------------------------------
E/ExoPlayerImplInternal: Source error.                                                                                 com.google.android.exoplayer2.source.UnrecognizedInputFormatException: None of the available extractors (MatroskaExtractor, FragmentedMp4Extractor, Mp4Extractor, Mp3Extractor, AdtsExtractor, Ac3Extractor, TsExtractor, FlvExtractor, OggExtractor, PsExtractor, WavExtractor) could read the stream.

-----------------------------------------------------------------------------------------------------

Then i run the app again and take this error.

-----------------------------------------------------------------------------------------------------

E/ExoPlayerImplInternal: Source error.
                                                                                         com.google.android.exoplayer2.ParserException: Searched too many bytes.
                                                                                             at com.google.android.exoplayer2.extractor.mp3.Mp3Extractor.synchronize(Mp3Extractor.java:265)
                                                                                             at com.google.android.exoplayer2.extractor.mp3.Mp3Extractor.read(Mp3Extractor.java:172)
                                                                                             at com.google.android.exoplayer2.source.ExtractorMediaPeriod$ExtractingLoadable.load(ExtractorMediaPeriod.java:714)
                                                                                             at com.google.android.exoplayer2.upstream.Loader$LoadTask.run(Loader.java:315)
                                                                                             at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1162)
                                                                                             at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:636)
                                                                                             at java.lang.Thread.run(Thread.java:764)

-----------------------------------------------------------------------------------------------------
And then i run one more time, this time it was working well but when i tried to seek i took above error one more time (source error).

**Comment by ojw28** [comment]
Please make sure you've erased the cache directory, since it may have incorrect data as a result of the previous bug in your code. The easiest way is probably to restart the emulator. Uninstalling may or may not have been sufficient, depending on where the cache directory was being placed on the device.

If you're still having issues then you'd probably be best off trying to start with our demo app, adding caching to that, and if you manage to reproduce the issue there providing us with some minimal reproduction steps. It's not really feasible for us to spend time debugging code that you've written from scratch.

Closing this for now because everything suggests this is an issue with your application code. Please let us know if you manage to reproduce the issue by adding to our demo app, with minimal reproduction steps. Thanks!

**Comment by ojw28** [comment]
Restarting an emulator apparently doesn't wipe data. There's a wipe data option if you right click on the virtual device in the Android Virtual Device Manager section of Android Studio.

**Comment by Toffor** [comment]
I fetched demo application and take a sample code from cache test and i found the problem. I debugged deeply and i discovered that i should provide a scratch space for AesChipperDataSink. 

I change this code block with, 

`        val aesCipherDataSink = AesCipherDataSink(secret, CacheDataSink(simpleCache, Long.MAX_VALUE))
`

this one

`val scratch = ByteArray(3897)
        val aesCipherDataSink = AesCipherDataSink(secret, CacheDataSink(simpleCache, Long.MAX_VALUE), scratch)
`

Now its working really fine. Maybe this can be helpful for others. Thank you very much!
